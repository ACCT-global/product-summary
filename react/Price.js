import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FormattedMessage, injectIntl, intlShape } from 'react-intl'

/**
 * The Price component. Shows the prices information of the Product Summary.
 */
class Price extends Component {
  static contextTypes = {
    culture: PropTypes.object,
  }

  static schema = {
    title: 'Price',
    description: 'The price component for the Product Summary',
    type: 'object',
    properties: {
      showListPrice: {
        type: 'boolean',
        title: 'Show list price',
      },
    },
  }

  render() {
    const {
      sellingPrice,
      listPrice,
      installments,
      installmentPrice,
      showListPrice,
      intl: { formatNumber },
    } = this.props

    const currencyOptions = {
      style: 'currency',
      currency: this.context.culture.currency,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }

    return (
      <div className="tc b fabriga">
        {showListPrice && (
          <div className="pv1">
            <div className="dib">
              <FormattedMessage id="pricing.from" />
            </div>
            <div className="dib strike ph2">
              {formatNumber(listPrice, currencyOptions)}
            </div>
          </div>
        )}
        <div className="pv1">
          <div className="dib">
            <FormattedMessage id="pricing.to" />
          </div>
          <div className="dib ph2">
            {formatNumber(sellingPrice, currencyOptions)}
          </div>
        </div>
        {installments &&
          installmentPrice && (
            <div>
              <div className="dib">
                <FormattedMessage
                  id="pricing.installment-display"
                  values={{
                    installments,
                    installmentPrice: formatNumber(
                      installmentPrice,
                      currencyOptions
                    ),
                  }}
                />
              </div>
            </div>
          )}
      </div>
    )
  }
}

Price.propTypes = {
  /** Product selling price */
  sellingPrice: PropTypes.number.isRequired,
  /** Product list price */
  listPrice: PropTypes.number.isRequired,
  /** Determines if the list price is shown or not */
  showListPrice: PropTypes.bool,
  /** Available number of installments */
  installments: PropTypes.number,
  /** Single installment price */
  installmentPrice: PropTypes.number,
  /** intl property to format data */
  intl: intlShape.isRequired,
}

export default injectIntl(Price)
