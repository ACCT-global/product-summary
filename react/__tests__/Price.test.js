/* eslint-env jest */
import React from 'react'
import { mountWithIntl, loadTranslation } from 'enzyme-react-intl'

import { getIntlContextInfo, getIntlInstance } from './helpers/intlHelper'
import Price from '../Price'

describe('<Price /> component', () => {
  let renderComponent, productPropsMock, configuration
  let intl, currencyOptions

  beforeEach(() => {
    const { context, childContextTypes, locale } = getIntlContextInfo()

    currencyOptions = {
      style: 'currency',
      currency: context.culture.currency,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }

    productPropsMock = {
      listPrice: 200,
      sellingPrice: 170,
      installments: 3,
      installmentPrice: 50,
    }
    configuration = {}

    intl = getIntlInstance()
    loadTranslation(`../locales/${locale}.json`)

    renderComponent = () => mountWithIntl(
      <Price {...productPropsMock} {...configuration} />, { context, childContextTypes }
    )
  })

  it('should be mounted and not break', () => {
    const component = renderComponent()
    expect(component).toMatchSnapshot()
  })

  it('should not show the list price if prop showListPrice is false', () => {
    configuration.showListPrice = false

    const component = renderComponent()

    expect(component.prop('showListPrice')).toBe(false)
    expect(component.contains(intl.formatMessage({ id: 'pricing.from' }))).toBe(false)
    expect(component.contains(intl.formatMessage({ id: 'pricing.to' }))).toBe(true)
    expect(component.contains(intl.formatNumber(productPropsMock.sellingPrice, currencyOptions))).toBe(true)
    expect(component.contains(intl.formatNumber(productPropsMock.listPrice, currencyOptions))).toBe(false)
  })

  describe('with no configuration', () => {
    it('should show the list price by default', () => {
      const component = renderComponent()
      expect(component.contains(intl.formatNumber(productPropsMock.listPrice, currencyOptions))).toBe(true)
    })

    it('should show the price labels by default', () => {
      const component = renderComponent()
      expect(component.contains(intl.formatMessage({ id: 'pricing.from' }))).toBe(true)
      expect(component.contains(intl.formatMessage({ id: 'pricing.to' }))).toBe(true)
    })

    it('should not show the installments by default', () => {
      const component = renderComponent()
      const formattedInstallmentPrice = intl.formatNumber(productPropsMock.installmentPrice, currencyOptions)
      const formattedMessage = intl.formatMessage({ id: 'pricing.installment-display' }, {
        installments: productPropsMock.installments,
        installmentPrice: formattedInstallmentPrice,
      })
      expect(component.contains(formattedMessage)).toBe(false)
    })
  })
})
