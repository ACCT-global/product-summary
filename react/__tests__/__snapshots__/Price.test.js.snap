// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Price /> component should be mounted and not break 1`] = `
<InjectIntl(Price)
  installmentPrice={50}
  installments={3}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {
        "loading": "Loading…",
        "pricing.from": "from:",
        "pricing.installment-display": "or up to {installments}X of {installmentPrice}",
        "pricing.to": "to:",
      },
      "now": [Function],
      "textComponent": "span",
    }
  }
  listPrice={200}
  sellingPrice={170}
>
  <Price
    installmentPrice={50}
    installments={3}
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "loading": "Loading…",
          "pricing.from": "from:",
          "pricing.installment-display": "or up to {installments}X of {installmentPrice}",
          "pricing.to": "to:",
        },
        "now": [Function],
        "textComponent": "span",
      }
    }
    listPrice={200}
    sellingPrice={170}
    showInstallments={false}
    showLabels={true}
    showListPrice={true}
  >
    <div
      className="tc b fabriga"
    >
      <div
        className="pv1"
      >
        <div
          className="dib"
        >
          <FormattedMessage
            id="pricing.from"
            values={Object {}}
          >
            <span>
              from:
            </span>
          </FormattedMessage>
        </div>
        <div
          className="dib strike ph2"
        >
          $200.00
        </div>
      </div>
      <div
        className="pv1"
      >
        <div
          className="dib"
        >
          <FormattedMessage
            id="pricing.to"
            values={Object {}}
          >
            <span>
              to:
            </span>
          </FormattedMessage>
        </div>
        <div
          className="dib ph2"
        >
          $170.00
        </div>
      </div>
    </div>
  </Price>
</InjectIntl(Price)>
`;
